worker-message/src/worker.rs

```diff
@@ -0,0 +1,325 @@
+        // Absolute path will replace final_path.
+        final_path.push(path)
+    }
+    final_path
+}
+
+async fn handle_request(req: Request, project_path: &Path) -> Result<Response> {
+    // let mut children = HashMap::<Pid, Child>::new();
+    match req {
+        Request::WriteFile(WriteFileRequest { path, content }) => {
+            let path = parse_working_dir(Some(path), project_path);
+            // Create intermediate directories.
+            if let Some(parent_dir) = path.parent() {
+                fs::create_dir_all(parent_dir).await?;
+            }
+            fs::write(path, content).await?;
+            Ok(Response::WriteFile(WriteFileResponse(())))
+        }
+        Request::ReadFile(ReadFileRequest { path }) => {
+            let path = parse_working_dir(Some(path), project_path);
+            let content = fs::read(path).await?;
+            Ok(Response::ReadFile(ReadFileResponse(content)))
+        }
+        Request::ExecuteCommand(ExecuteCommandRequest {
+            cmd,
+            args,
+            envs,
+            cwd,
+        }) => {
+            let Output {
+                status,
+                stderr,
+                stdout,
+            } = Command::new(cmd)
+                .args(args)
+                .envs(envs)
+                .current_dir(parse_working_dir(cwd, project_path))
+                .kill_on_drop(true)
+                .output()
+                .await?;
+            Ok(Response::ExecuteCommand(ExecuteCommandResponse(
+                ExecuteOutput {
+                    status: status.code(),
+                    stderr,
+                    stdout,
+                },
+            )))
+        }
+        _ => unreachable!(),
+    }
+}
+
+async fn stream_stdio(
+    coordinator_tx: mpsc::Sender<WorkerMessage>,
+    mut stdin_rx: mpsc::Receiver<String>,
+    child: &mut Child,
+) -> Result<()> {
+    let mut stdin = child
+        .stdin
+        .take()
+        .ok_or(Error::msg("child process's stdin is not captured"))?;
+    let stdout = child
+        .stdout
+        .take()
+        .ok_or(Error::msg("child process's stdout is not captured"))?;
+    let stderr = child
+        .stderr
+        .take()
+        .ok_or(Error::msg("child process's stderr is not captured"))?;
+
+    let mut set = JoinSet::new();
+
+    set.spawn(async move {
+        loop {
+            let data = stdin_rx
+                .recv()
+                .await
+                .ok_or(Error::msg("user input channel closed"))?;
+            stdin.write_all(data.as_bytes()).await?;
+            stdin.flush().await?;
+        }
+        #[allow(unreachable_code)]
+        Ok::<(), Error>(())
+    });
+    let coordinator_tx_out = coordinator_tx.clone();
+    set.spawn(async move {
+        let mut stdout_buf = BufReader::new(stdout);
+        loop {
+            // Must be valid UTF-8.
+            let mut buffer = String::new();
+            let n = stdout_buf.read_line(&mut buffer).await?;
```

Comment:
I choose `read_line()` because I am worried that bytes read might be partial UTF-8. 
But with more thinking, it's recoverable with `FromUtf8Error`.
As for non-UTF-8 data, it adds some complexity of sending to frontend. (JSON cannot contain binary).

Reply:
Yes, `read_line` is the correct choice to start with, and partial UTF-8 bytes is exactly the reason. And yes, `FromUtf8Error` would allow finding the last valid UTF-8 data and sending that that, keeping the partial bytes for the next read.

Non-UTF-8 would indeed need to be encoded in another way. I don't have any specific ideas, and it may never happen (so we shouldn't worry about it now), but we could do something easy like base64 encode it or something complicated like using a `Message::Binary` for that specific data.